File: ios/StackSpeak/Services/Backend/NoOpBackendService.swift
import Foundation

/// Fallback used when no backend is configured (no `Supabase.plist`, or a free
/// user with no cross-platform sync). Reports unconfigured and no-ops every
/// call, so the rest of the app can depend on `BackendService` unconditionally
/// without branching on "is sync available".
struct NoOpBackendService: BackendService {
    var isConfigured: Bool { false }

    func ensureSession() async throws -> BackendUserID {
        throw BackendError.notConfigured
    }

    func signInWithApple(idToken: String, rawNonce: String) async throws -> BackendUserID {
        throw BackendError.notConfigured
    }

    func signUpWithEmail(_ email: String, password: String) async throws -> EmailSignUpResult {
        throw BackendError.notConfigured
    }

    func signInWithEmail(_ email: String, password: String) async throws -> BackendUserID {
        throw BackendError.notConfigured
    }

    func sendPasswordReset(email: String) async throws {
        throw BackendError.notConfigured
    }

    func fetchSnapshot() async throws -> ProgressSnapshot? {
        throw BackendError.notConfigured
    }

    func pushSnapshot(_ snapshot: ProgressSnapshot) async throws {
        throw BackendError.notConfigured
    }

    func signOut() async {}
}

File: android/app/src/main/java/com/stackspeak/data/backend/NoOpBackendService.kt
package com.stackspeak.data.backend

import com.stackspeak.domain.ProgressSnapshot

/** Null-object backend for when no Supabase config is present (local-only build). */
class NoOpBackendService : BackendService {
    override val isConfigured: Boolean = false
    override suspend fun ensureSession(): BackendUserId = throw BackendError.NotConfigured
    override suspend fun signUpWithEmail(email: String, password: String): EmailSignUpResult = throw BackendError.NotConfigured
    override suspend fun signInWithEmail(email: String, password: String): BackendUserId = throw BackendError.NotConfigured
    override suspend fun sendPasswordReset(email: String) = throw BackendError.NotConfigured
    override suspend fun signInWithGoogle(present: WebAuthPresenter): BackendUserId = throw BackendError.NotConfigured
    override suspend fun fetchSnapshot(): ProgressSnapshot? = throw BackendError.NotConfigured
    override suspend fun pushSnapshot(snapshot: ProgressSnapshot) = throw BackendError.NotConfigured
    override suspend fun signOut() {}
}
